### DISCUSSION TITLE
Distributions beyond sympy

### DISCUSSION THREAD

**Discussion by jwnys** [maichmueller/scm#8] [Q&A]
I'm continuing the discussion from a closed issue here:
What it the reason for using sympy?
Would it also be possible to wrap distributions from scipy or custom numpy distributions?

Best, Jannes

**Comment by maichmueller** [comment]
Hi @jwnys, I chose to use SymPy specifically after attempting the creation of my own noise generators and Assignment parsing system. As it turns out, it is simply too much work (at least at the moment) to provide my own mini-CAS system to generate arbitrary functions and noise distributions.

The benefits of sympy are then quite a few (feel free to add to this):
- An entire CAS system to build on
- Simple creation of functions through string descriptions: This allows a user to state relatively complex functions with ease. No need to think of which library or functions to import, just make it a string and sympy should be able to parse it. AFAIK SymPy parses with the notion of substituting as many function strings with numpy methods as possible, so that vectorization for most elementary methods is not an issue.
- Sympy supports (custom) operations on random variables, i.e. applying a function on one (or more) RV and still being able to sample, compute the expectation etc. (This is not currently well supported in scmodels. I'm working on it.) This functionality allows a user to set a complex noise variable, such as sqrt(log(Normal) ** 2) with relative ease (no need to build a custom distribution for this).
- Sympy supports Scipy, numpy and theano distributions. So long as you have these packages installed, sympy can use them. The list of wrapped distributions is also quite large, so I don't see the need for change in this regard atm. Due to seeding however, scmodels only supports scipy and numpy really. There is no direct check on this however.
- For custom distributions sympy offers FiniteDistributionHandmade, ContinuousDistributionHandmade where you can provide the pdf. For more questions on that I'd recommend asking the sympy developers.  

The downsides as I see them are the following (feel free to add to this):
- Sympy dependency and potential limitations by the project (e.g. seeding of RVs was not possible before I pushed for it).
- The need to refer a user to SymPy's API, whose documentation can be so-so at times.

Hope this answers your question.
